core/node/vm_runner/src/output_handler.rs

```diff
@@ -0,0 +1,541 @@
+use std::{
+    fmt::{Debug, Formatter},
+    sync::Arc,
+    time::Duration,
+};
+
+use async_trait::async_trait;
+use dashmap::DashMap;
+use futures::future::BoxFuture;
+use tokio::sync::{oneshot, watch};
+use zksync_core::state_keeper::{StateKeeperOutputHandler, UpdatesManager};
+use zksync_dal::{ConnectionPool, Core};
+use zksync_types::L1BatchNumber;
+
+use crate::VmRunnerIo;
+
+/// Functionality to produce a [`StateKeeperOutputHandler`] implementation for a specific L1 batch.
+///
+/// The idea behind this trait is that often handling output data is independent of the order of the
+/// batch that data belongs to. In other words, one could be handling output of batch #100 and #1000
+/// simultaneously. Implementing this trait signifies that this property is held for the data the
+/// implementation is responsible for.
+#[async_trait]
+pub trait OutputHandlerFactory: Debug + Send {
+    /// Creates a [`StateKeeperOutputHandler`] implementation for the provided L1 batch. Only
+    /// supposed to be used for the L1 batch data it was created against. Using it for anything else
+    /// is undefined behavior.
+    ///
+    /// # Errors
+    ///
+    /// Propagates DB errors.
+    async fn create_handler(
+        &mut self,
+        l1_batch_number: L1BatchNumber,
+    ) -> anyhow::Result<Box<dyn StateKeeperOutputHandler>>;
+}
+
+/// A delegator factory that requires an underlying factory `F` that does the actual work, however
+/// this struct is orchestrated such that any output handler it produces has a non-blocking
+/// `handle_l1_batch` implementation (where the heaviest work is expected to happen).
+///
+/// Once the asynchronous work done in `handle_l1_batch` finishes it is also guaranteed to mark the
+/// batch is processed by `Io`. It is guaranteed, however, that for any processed batch all batches
+/// preceding it are also processed. No guarantees about subsequent batches. For example, if
+/// batches #1, #2, #3, #5, #9, #100 are processed then only batches #{1-3} will be marked as
+/// processed and #3 would be the latest processed batch as defined in [`VmRunnerIo`].
+pub struct ConcurrentOutputHandlerFactory<Io: VmRunnerIo, F: OutputHandlerFactory> {
+    pool: ConnectionPool<Core>,
+    state: Arc<DashMap<L1BatchNumber, oneshot::Receiver<BoxFuture<'static, anyhow::Result<()>>>>>,
```

Comment:
- Nit: Could make sense to define a type alias for `oneshot::Receiver<_>` so that it's easier to read.
- Is using a concurrent map (vs `Mutex<HashMap<_>>`) really warranted here? As long as we don't need references from the map, looks like a `Mutex` would do fine.

Reply:
> Is using a concurrent map (vs Mutex<HashMap<_>>) really warranted here? As long as we don't need references from the map, looks like a Mutex would do fine.

Just seemed natural to use a concurrent map given that we never touch the same entries at the same time. Also, `dashmap` is already a transitive dependency in the system so just felt it would be appropriate to depend on it.

Reply:
Resolving this but feel free to reopen